### PR TITLE
fix(sql): breaking💥 fix coalesce function correctness

### PR DESCRIPTION
**Pull request by bluestreak01**
fixes https://github.com/questdb/questdb/issues/4478
fixes https://github.com/questdb/questdb/issues/4477
fixes https://github.com/questdb/questdb/issues/4505

### Scope

This PR impacts all SQL functions that accept var-arg argument. More specifically impacted functions are:
- COALESCE
- IN
- CASE
- CAST(NULL AS...

### Bind variable behaviour (general)

Bind variables are utilized by:
- preparing SQL statement
- setting bind variable value and type
- execute SQL to retrieve result for the bind variable value

The prepared statement can be reused. Potentially allowing user to change bind variable type from one execution to the next.

### Old behaviour

The affected functions use to accept **bind variables** and implicitly type them as `STRING` during the preparation stage. When user sets type of bind variable as string, the result would be predictable. However, if user sets bind variable type to something other than string - the result is generally undefined. You can get anything between incorrect result and an obscure error.

### New behaviour

After SQL preparation stage the type of bind variables is **UNDEFINED**. The type becomes defined once bind variable is assigned a value. Some functions cannot allow that, because type of their arguments also influences the return type of the function itself. As a result these function no longer accept bind variables:

- COALESCE
- CASE

Additionally this function:

- COALESCE

enforces the types of its arguments to be "compatible", e.g. to belong to the same group of types. The groups of types are:

- STRING,VARCHAR,CHAR
- DOUBLE, FLOAT, INT, LONG, SHORT, BYTE
- BOOLEAN
- UUID
- LONG256
- IPV4

arguments, which types belong to different groups will be rejected by `COALESCE

This function:

- IN

will now narrow down the type of bind variables to those compatible with the type of its left argument. The types must belong to the same group, as described for `COALESCE`

### CODE REVIEW

**Comment by jerrinot** [comment]
this should be marked as a _breaking change💥_

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 129 / 157 (82.17%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|54|63|85.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 126 / 154 (81.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|53|62|85.48%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 135 / 160 (84.38%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|68|91.18%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 174 / 202 (86.14%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|68|91.18%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|39|42|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 266 / 301 (88.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InDoubleFunctionFactory.java|40|46|86.96%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|69|89.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|39|42|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 298 / 337 (88.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InLongFunctionFactory.java|30|35|85.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|69|89.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InDoubleFunctionFactory.java|42|47|89.36%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|39|42|92.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 540 / 592 (91.22%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InLongFunctionFactory.java|35|40|87.50%|
|:large_blue_circle:|io/questdb/std/Uuid.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/std/Numbers.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|69|89.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InDoubleFunctionFactory.java|47|52|90.38%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InUuidFunctionFactory.java|68|74|91.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ConcatFunctionFactory.java|51|53|96.23%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndUUIDCFunctionFactory.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InCharFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/LongLongHashSet.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|3|3|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java [hunk 1]
```diff
@@ -138,9 +141,13 @@ public int define(int index, int type, int position) throws SqlException {
                 return type;
             case ColumnType.STRING:
             case ColumnType.SYMBOL:
-            case ColumnType.VAR_ARG:
                 setStr(index);
                 return ColumnType.STRING;
+            case ColumnType.VAR_ARG:
+                // we cannot define bind variable as vararg, it is
+                // a code for method signature and is not a "type"
+                assert false;
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java:149]: this should never happen in real life and cannot be tested

File: core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java [hunk 2]
```diff
@@ -10087,6 +10123,9 @@ private void assertQueryAgainstIndexedSymbol(String[] values, String whereClause
             }
             metaSink.put("\nts option: ").put(tsOption);
 
+            if (iter == 5) {
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java]: cleaned in [f059980](https://github.com/questdb/questdb/pull/4490/commits/f059980d7fd6267588edcda200a71bd38c9c4892)

File: core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java [hunk 3]
```diff
@@ -10076,6 +10110,8 @@ private void assertQueryAgainstIndexedSymbol(String[] values, String whereClause
 
             String query = "select s from tab where " + where;
 
+            System.out.println("iter: " + iter + ", sql:" + query);
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/test/java/io/questdb/test/cutlass/pgwire/PGJobContextTest.java]: done in [f059980](https://github.com/questdb/questdb/pull/4490/commits/f059980d7fd6267588edcda200a71bd38c9c4892)

File: core/src/main/java/io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java]: replaced with SqlException in [f059980](https://github.com/questdb/questdb/pull/4490/commits/f059980d7fd6267588edcda200a71bd38c9c4892)

File: core/src/main/java/io/questdb/griffin/FunctionParser.java [hunk 4]
```diff
@@ -397,7 +397,7 @@ private Function checkAndCreateFunction(
         final int position = node.position;
         Function function;
         try {
-            LOG.debug().$("call ").$(node).$(" -> ").$(factory.getSignature()).$();
+            LOG.info().$("call ").$(node).$(" -> ").$(factory.getSignature()).$("[factory=").$(factory).$(']').$();
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/FunctionParser.java]: cleaned up in [f059980](https://github.com/questdb/questdb/pull/4490/commits/f059980d7fd6267588edcda200a71bd38c9c4892)

File: core/src/main/java/io/questdb/griffin/engine/functions/conditional/NullCaseFunction.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    @Override
+    public CharSequence getSymbolB(Record rec) {
+        return null;
+    }
+
+    @Override
+    public long getTimestamp(Record rec) {
+        return 0L;
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/conditional/NullCaseFunction.java]: done in https://github.com/questdb/questdb/pull/4490/commits/b148482b5d983bdc0902100b6a89a09ed00da100

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 745 / 846 (88.06%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/NullCaseFunction.java|5|44|11.36%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|5|40.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|44|52|84.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InLongFunctionFactory.java|35|40|87.50%|
|:large_blue_circle:|io/questdb/std/Uuid.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/std/Numbers.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|69|89.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InDoubleFunctionFactory.java|47|52|90.38%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InUuidFunctionFactory.java|68|74|91.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ConcatFunctionFactory.java|51|53|96.23%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|154|155|99.35%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndUUIDCFunctionFactory.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InCharFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/LongLongHashSet.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 745 / 846 (88.06%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/NullCaseFunction.java|5|44|11.36%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|5|40.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|42|59|71.19%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|44|52|84.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InLongFunctionFactory.java|35|40|87.50%|
|:large_blue_circle:|io/questdb/std/Uuid.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/std/Numbers.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|69|89.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InDoubleFunctionFactory.java|47|52|90.38%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InUuidFunctionFactory.java|68|74|91.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ConcatFunctionFactory.java|51|53|96.23%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|154|155|99.35%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndUUIDCFunctionFactory.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InCharFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/LongLongHashSet.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Review by ideoma — APPROVED** [review]

File: core/src/test/java/io/questdb/test/griffin/WhereClauseParserTest.java [hunk 6]
```diff
@@ -2362,6 +2364,9 @@ public void testIntervalVarchar() throws Exception {
                 "[]"
         );
 
+        System.out.println(Timestamps.toString(1648202400000000L));
+        System.out.println(Timestamps.toString(1648166399999999L));
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/test/java/io/questdb/test/griffin/WhereClauseParserTest.java]: removed in https://github.com/questdb/questdb/pull/4490/commits/746da8780c80fe054995e524f387f06af6e56374

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 771 / 872 (88.42%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlException.java|0|3|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/NullCaseFunction.java|5|44|11.36%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/BindVariableServiceImpl.java|2|5|40.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|68|85|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CoalesceFunctionFactory.java|44|52|84.62%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InLongFunctionFactory.java|35|40|87.50%|
|:large_blue_circle:|io/questdb/std/Uuid.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/std/Numbers.java|8|9|88.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampTimestampFunctionFactory.java|62|69|89.86%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InDoubleFunctionFactory.java|47|52|90.38%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InUuidFunctionFactory.java|68|74|91.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InVarcharFunctionFactory.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/str/ConcatFunctionFactory.java|51|53|96.23%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseCommon.java|154|155|99.35%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/CaseFunctionFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InTimestampStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/finance/LevelTwoPriceFunctionFactory.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/ExplainPlanFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/conditional/SwitchFunctionFactory.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/rnd/RndUUIDCFunctionFactory.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bind/IndexedParameterLinkFunction.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|14|14|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/Constants.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/model/IntervalUtils.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InCharFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlExecutionContextImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/catalogue/TypeOfFunctionFactory.java|31|31|100.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/LongLongHashSet.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|4|4|100.00%|

**Pull request merged**